Fix serial IRQ handling

fkjagodzinski opened the pull request:
### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
Check that the RX or TX interrupt is enabled before calling a registered handler with `RxIrq` or `TxIrq` arg.

Fixes #10773

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

fkjagodzinski committed:
K64F: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
STM32F7: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

LMESTM reviewed (commented):
Looks good to me - thanks.
should be applied to all STM32 targets in a similar way (the same MACROs should be available)

fkjagodzinski committed:
STM32F0: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
STM32L0: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
STM32F3: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
STM32H7: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
STM32L4: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
STM32WB: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
K22F: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
K24F: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
KW41Z: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
KW24D: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
KL82Z: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
KL43Z: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
K66F: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
KL27Z: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

fkjagodzinski committed:
K82F: Fix serial IRQ handling

Check that the RX or TX interrupt is enabled before calling
a registered handler with RxIrq or TxIrq arg.

LMESTM commented on targets/TARGET_STM/TARGET_STM32F0/serial_device.c:
```diff
@@ -62,12 +62,12 @@ static void uart_irq(UARTName uart_name)
         if (serial_irq_ids[id] != 0) {
             if (__HAL_UART_GET_FLAG(huart, UART_FLAG_TXE) != RESET) {
-                if (__HAL_UART_GET_IT(huart, UART_IT_TXE) != RESET) {
+                if (__HAL_UART_GET_IT(huart, UART_IT_TXE) != RESET && __HAL_UART_GET_IT_SOURCE(huart, UART_IT_TXE)) {
```
Ok so in case Interrupt has been disabled but FLAG is on anyway, handler will not be called, then shouldn't we clear the flag ?

LMESTM replied:
Giving it a second thought ... that should not be required. The flag could be left in current state I guess until inerrupt is activated again

fkjagodzinski replied:
Good point; I haven't thought about that. In a normal flow, these flags are cleared by the handler (provided it actually calls `serial_putc()` / `serial_getc()`):
* TXE is cleared by a write to the USART_TDR register,
* RXNE is cleared by a read to the USART_RDR register.

When the handler is not called, the flags will remain asserted. I can't think of any side effects of that right now. Both flags may be cleared with the use of USART_RQR though.

LMESTM reviewed (commented):
Just one question left to be answered

LMESTM reviewed (approved):
for STM targets - approved
